sqlmesh/core/test/definition.py

```diff
@@ -310,17 +314,38 @@ def _validate_and_normalize_test(self) -> None:
         partial = outputs.pop("partial", None)
 
         def _normalize_rows(
-            values: t.List[Row] | t.Dict, name: str, partial: bool = False
+            values: t.List[Row] | t.Dict,
+            name: str,
+            partial: bool = False,
+            dialect: DialectType = None,
         ) -> t.Dict:
             if not isinstance(values, dict):
                 values = {"rows": values}
 
-            if "rows" not in values:
+            rows = values.get("rows")
+            query = values.get("query")
+
+            if query is not None:
+                if rows is not None:
+                    _raise_error(
+                        f"Invalid test, cannot set both 'query' and 'rows' for '{name}'", self.path
+                    )
+
+                # We parse the user-supplied query using the testing adapter dialect, but we
+                # normalize its identifiers according to the model's dialect, so that, e.g.,
+                # the projection names match those in its `columns_to_types` field
+                values["query"] = normalize_identifiers(
```

Comment:
this is confusing, but i think it makes sense, we need to normalize by the model, that's always true or else references will be wrong.

the parsing should be supplied by the actual test engine so there are no nuanences with transpilation

Reply:
both points are exactly what I had in mind 👍